File: GraffiTab-iOS/Classes/ViewControllers/Main/Main/Locations/CreateLocationViewController.m
//
//  CreateLocationViewController.m
//  GraffiTab-iOS
//
//  Created by Georgi Christov on 11/12/2014.
//  Copyright (c) 2014 GraffiTab. All rights reserved.
//

#import "CreateLocationViewController.h"
#import <AddressBookUI/AddressBookUI.h>
#import "BlockActionSheet.h"

@interface CreateLocationViewController () {
    
    IBOutlet UIImageView *mapCenterImage;
    IBOutlet MKMapView *mapView;
    IBOutlet UIButton *backButton;
    IBOutlet UIButton *searchButton;
    IBOutlet UIButton *gridButton;
    IBOutlet UIButton *locateButton;
    IBOutlet UIView *segmentView;
    IBOutlet UIView *searchView;
    IBOutlet UITextField *searchField;
    
    BOOL showedFirstUserLocation;
    BOOL isSearching;
    CLPlacemark *lastPlacemark;
}

- (IBAction)onClickBack:(id)sender;
- (IBAction)onClickSearch:(id)sender;
- (IBAction)onClickCreate:(id)sender;
- (IBAction)onClickLocate:(id)sender;

@end

@implementation CreateLocationViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    [self setupButtons];
    [self setupImageViews];
    [self setupMapView];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    [[UIApplication sharedApplication] setStatusBarStyle:UIStatusBarStyleDefault animated:YES];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    [[UIApplication sharedApplication] setStatusBarStyle:UIStatusBarStyleLightContent animated:YES];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)dealloc {
#ifdef DEBUG
    NSLog(@"DEALLOC %@", self.class);
#endif
    
    switch (mapView.mapType) {
        case MKMapTypeHybrid:
            mapView.mapType = MKMapTypeStandard;
            break;
        case MKMapTypeStandard:
            mapView.mapType = MKMapTypeHybrid;
            break;
        default:
            break;
    }
    
    mapView.showsUserLocation = NO;
    mapView.delegate = nil;
    [mapView removeFromSuperview];
    mapView = nil;
}

- (IBAction)onClickBack:(id)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (IBAction)onClickSearch:(id)sender {
    isSearching = !isSearching;
    
    if (isSearching) {
        [UIView animateWithDuration:0.5 animations:^{
            CGRect f = searchView.frame;
            f.origin.x = backButton.frame.origin.x + backButton.frame.size.width + 30;
            f.size.width = self.view.frame.size.width - f.origin.x - 10;
            searchView.frame = f;
            
            searchField.alpha = 1.0;
        } completion:^(BOOL finished) {
            [searchField becomeFirstResponder];
        }];
    }
    else {
        [self.view endEditing:YES];
        
        [UIView animateWithDuration:0.5 animations:^{
            CGRect f = searchView.frame;
            f.size.width = backButton.frame.size.width;
            f.origin.x = self.view.frame.size.width - 10 - f.size.width;
            searchView.frame = f;
            
            searchField.alpha = 0.0;
        }];
    }
}

- (IBAction)onClickCreate:(id)sender {
    if (!lastPlacemark)
        [Utils showMessage:APP_NAME message:@"Please select a location first."];
    else {
        [[LoadingViewManager getInstance] addLoadingToView:self.view withMessage:@"Processing..."];
        
        [GTLocationManager createLocationWithPlacemark:lastPlacemark successBlock:^(GTResponseObject *response) {
            [[LoadingViewManager getInstance] removeLoadingView];
            
            [Utils showMessage:APP_NAME message:@"Your location was created successfully."];
            
            [[NSNotificationCenter defaultCenter] postNotificationName:NOTIFICATION_UPDATE_LOCATIONS object:nil];
            
            // Delay execution of my block for 10 seconds.
            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.3 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
                [self onClickBack:nil];
            });
        } failureBlock:^(GTResponseObject *response) {
            [[LoadingViewManager getInstance] removeLoadingView];
            
            if (response.reason == AUTHORIZATION_NEEDED) {
                [Utils logoutUserAndShowLoginController];
                [Utils showMessage:APP_NAME message:@"Your session has timed out. Please login again."];
            }
            else if (response.reason == FORBIDDEN)
                [Utils showMessage:APP_NAME message:@"You have reached the maximum number of locations."];
            else
                [Utils showMessage:APP_NAME message:response.message];
        }];
    }
}

- (IBAction)onClickLocate:(id)sender {
    [self centerToLocation:mapView.userLocation.location];
}

- (void)centerToLocation:(CLLocation *)location {
    [self zoomMapToLocation:location];
}

#pragma mark - Searching

- (void)searchLocationForAddress:(NSString *)address {
    [[LoadingViewManager getInstance] addLoadingToView:self.view withMessage:@"Processing..."];
    
    MKLocalSearchRequest* request = [[MKLocalSearchRequest alloc] init];
    request.naturalLanguageQuery = address;
    
    MKLocalSearch *search = [[MKLocalSearch alloc] initWithRequest:request];
    [search startWithCompletionHandler:^(MKLocalSearchResponse *response, NSError *error) {
        NSArray *placemarks = response.mapItems; // array of MKMapItems
        
        [[LoadingViewManager getInstance] removeLoadingView];
        
        if (placemarks.count <= 0) {
            [Utils showMessage:APP_NAME message:@"No locations found for this address."];
            return;
        }
        
        if (placemarks.count > 1) {
            // More than one matches found, so ask user which one to use.
            
            BlockActionSheet *actionSheet = [[BlockActionSheet alloc] initWithTitle:@"Select address"];
            
            for (MKMapItem *mapItem in placemarks) {
                // Process the placemark.
                NSString *title = ABCreateStringWithAddressDictionary(mapItem.placemark.addressDictionary, YES);
                
                [actionSheet addButtonWithTitle:ABCreateStringWithAddressDictionary(mapItem.placemark.addressDictionary, NO) block:^{
                    searchField.text = title;
                    
                    [self zoomMapToLocation:mapItem.placemark.location];
                }];
            }
            
            [actionSheet addButtonWithTitle:@"Cancel" block:nil];
            [actionSheet showInView:self.view];
        }
        else {
            // Only one match found, so use it.
            MKMapItem *mapItem = [placemarks objectAtIndex:0];
            
            [self zoomMapToLocation:mapItem.placemark.location];
        }
    }];
}

- (void)zoomMapToLocation:(CLLocation *)location {
    MKCoordinateRegion mapRegion;
    mapRegion.center = location.coordinate;
    mapRegion.span.latitudeDelta = 0.2;
    mapRegion.span.longitudeDelta = 0.2;
    
    [mapView setRegion:mapRegion animated:YES];
}

#pragma mark - UITextFieldDelegate

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    [textField resignFirstResponder];
    
    if (textField.text.length > 0)
        [self searchLocationForAddress:textField.text];
    
    return YES;
}

#pragma mark - MKMapViewDelegate

- (void)mapView:(MKMapView *)mv didUpdateUserLocation:(MKUserLocation *)userLocation {
    if (!showedFirstUserLocation)
        [self centerToLocation:mapView.userLocation.location];
    
    showedFirstUserLocation = YES;
}

- (void)mapView:(MKMapView *)mv regionDidChangeAnimated:(BOOL)animated {
    CLGeocoder *geocoder = [[CLGeocoder alloc] init];
    [geocoder reverseGeocodeLocation:[[CLLocation alloc] initWithLatitude:mapView.centerCoordinate.latitude longitude:mapView.centerCoordinate.longitude]
                   completionHandler:^(NSArray *placemarks, NSError *error) {
                       if (error){
                           NSLog(@"Geocode failed with error: %@", error);
                           return;
                       }
                       
                       if (placemarks && placemarks.count > 0) {
                           lastPlacemark = [placemarks objectAtIndex:0];
                           
                           NSString *addressTxt = ABCreateStringWithAddressDictionary(lastPlacemark.addressDictionary, NO);
                           NSString *newString = [[addressTxt componentsSeparatedByCharactersInSet:[NSCharacterSet newlineCharacterSet]] componentsJoinedByString:@", "];
                           searchField.text = newString;
                       }
                   }];
}

- (void)centerToUserLocation {
    MKCoordinateRegion mapRegion;
    mapRegion.center = mapView.userLocation.coordinate;
    mapRegion.span.latitudeDelta = 0.2;
    mapRegion.span.longitudeDelta = 0.2;
    
    [mapView setRegion:mapRegion animated: YES];
}

#pragma mark - Setup

- (void)setupMapView {
    mapView.rotateEnabled = NO;
}

- (void)setupImageViews {
    CGRect f = mapCenterImage.frame;
    f.origin.y = mapView.center.y - f.size.height / 2 - 10;
    mapCenterImage.frame = f;
}

- (void)setupButtons {
    NSArray *views = @[backButton, searchView, segmentView];
    
    for (UIButton *b in views) {
        b.layer.cornerRadius = 5.0f;
        b.layer.masksToBounds = NO;
        b.layer.shadowColor = [UIColor blackColor].CGColor;
        b.layer.shadowOpacity = 0.5;
        b.layer.shadowRadius = 1;
        b.layer.shadowOffset = CGSizeMake(0.5, 0.5);
    }
    
    [backButton setImage:[backButton.imageView.image imageWithTint:UIColorFromRGB(COLOR_MAIN)] forState:UIControlStateNormal];
    [searchButton setImage:[searchButton.imageView.image imageWithTint:UIColorFromRGB(COLOR_MAIN)] forState:UIControlStateNormal];
    [gridButton setImage:[gridButton.imageView.image imageWithTint:UIColorFromRGB(COLOR_MAIN)] forState:UIControlStateNormal];
    [locateButton setImage:[locateButton.imageView.image imageWithTint:UIColorFromRGB(COLOR_MAIN)] forState:UIControlStateNormal];
}

@end

File: GraffiTab/Classes/ViewControllers/Home/Me/Locations/CreateLocationViewController.swift
//
//  CreateLocationViewController.swift
//  GraffiTab
//
//  Created by Georgi Christov on 23/05/2016.
//  Copyright © 2016 GraffiTab. All rights reserved.
//

import UIKit
import MapKit
import GraffiTab_iOS_SDK
import AddressBookUI
import CocoaLumberjack

class CreateLocationViewController: BackButtonViewController, UITextFieldDelegate, MKMapViewDelegate {
    
    @IBOutlet weak var mapView: MKMapView!
    @IBOutlet weak var searchField: UITextField!
    @IBOutlet weak var backBtn: TintButton!
    @IBOutlet weak var searchBtn: TintButton!
    @IBOutlet weak var searchContainer: UIView!
    @IBOutlet weak var searchWidthConstraint: NSLayoutConstraint!
    @IBOutlet weak var loadingIndicator: UIActivityIndicatorView!
    @IBOutlet weak var createBtn: MaterializeRoundButton!
    @IBOutlet weak var mapCenterImage: TintImageView!
    
    var toEdit: GTLocation?
    
    var lastPlacemark: CLPlacemark?
    var lastPlacemarkAddress: String?
    var isMovedByTap = false
    var isSearching = false
    var showedFirstUserLocation = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Do any additional setup after loading the view.
        
        setupButtons()
        setupMapView()
        setupImageViews()
        
        if toEdit != nil {
            centerToLocation(CLLocation(latitude: toEdit!.latitude!, longitude: toEdit!.longitude!))
        }
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        // Register analytics events.
        AnalyticsUtils.sendScreenEvent(self)
        
        if self.navigationController != nil && !self.navigationController!.navigationBarHidden {
            self.navigationController?.setNavigationBarHidden(true, animated: true)
        }
    }
    
    @IBAction func onClickCreate(sender: AnyObject) {
        DDLogDebug("[\(NSStringFromClass(self.dynamicType))] Attempting to create location")
        
        if lastPlacemark == nil {
            DialogBuilder.showErrorAlert(self, status: NSLocalizedString("controller_create_location_not_selected", comment: ""), title: App.Title)
        }
        else {
            // Register analytics events.
            AnalyticsUtils.sendAppEvent("location_create", label: nil)
            
            self.view.showActivityViewWithLabel(NSLocalizedString("other_processing", comment: ""))
            self.view.rn_activityView.dimBackground = false
            
            let success = {
                self.view.hideActivityView()
                
                DialogBuilder.showSuccessAlert(self, status: NSLocalizedString("controller_create_location_saved", comment: ""), title: App.Title, okAction: {
                    Utils.runWithDelay(0.3, block: {
                        self.onClickBack(nil)
                    })
                })
            }
            let failure = {(response: GTResponseObject) in
                self.view.hideActivityView()
                
                DialogBuilder.showAPIErrorAlert(self, status: response.error.localizedMessage(), title: App.Title, forceShow: true, reason: response.error.reason)
            }
            
            if toEdit != nil {
                GTMeManager.editLocation(toEdit!.id!, address: lastPlacemarkAddress!, latitude: lastPlacemark!.location!.coordinate.latitude, longitude: lastPlacemark!.location!.coordinate.longitude, successBlock: { (response) in
                    success()
                    }, failureBlock: { (response) in
                        failure(response)
                })
            }
            else {
                GTMeManager.createLocation(lastPlacemarkAddress!, latitude: lastPlacemark!.location!.coordinate.latitude, longitude: lastPlacemark!.location!.coordinate.longitude, successBlock: { (response) in
                    success()
                }, failureBlock: { (response) in
                    failure(response)
                })
            }
        }
    }
    
    @IBAction func onClickBack(sender: AnyObject?) {
        let destroy = {
            self.mapView.mapType = self.mapView.mapType == .Standard ? .Satellite : .Standard
            self.mapView.delegate = nil
            self.mapView.removeFromSuperview()
            self.mapView = nil
        }
        
        if self.navigationController?.viewControllers.count <= 1 { // We're running in a container so show a close button.
            self.dismissViewControllerAnimated(true, completion: {
                destroy()
            })
        }
        else {
            self.navigationController?.popViewControllerAnimated(true)
            
            Utils.runWithDelay(0.3, block: destroy)
        }
    }
    
    @IBAction func onClickSearch(sender: AnyObject) {
        DDLogDebug("[\(NSStringFromClass(self.dynamicType))] Attempting to search for location")
        
        isSearching = !isSearching
        
        let searchWidth: CGFloat
        if isSearching {
            searchWidth = 220
        }
        else {
            searchWidth = 40
        }
        
        searchWidthConstraint.constant = searchWidth
        searchContainer.setNeedsUpdateConstraints()
        UIView.animateWithDuration(0.3, animations: {
            self.searchContainer.layoutIfNeeded()
        }) { (finished) in
            if finished {
                if self.isSearching {
                    self.searchField.becomeFirstResponder()
                }
                else {
                    self.view.endEditing(true)
                }
            }
        }
    }
    
    // MARK: - Search
    
    func searchLocationForAddress(address: String) {
        self.view.showActivityViewWithLabel(NSLocalizedString("other_processing", comment: ""))
        self.view.rn_activityView.dimBackground = false
        
        let request = MKLocalSearchRequest()
        request.naturalLanguageQuery = address
        
        let search = MKLocalSearch(request: request)
        search.startWithCompletionHandler { (response, error) in
            self.view.hideActivityView()
            
            let placemarks = response?.mapItems
            if placemarks?.count <= 0 {
                DialogBuilder.showOKAlert(self, status: NSLocalizedString("controller_create_location_none", comment: ""), title: App.Title)
            }
            
            if placemarks?.count > 1 { // More than 1 address matches found. Ask user which one to use.
                DialogBuilder.showOKAlert(self, status: NSLocalizedString("controller_create_location_multiple", comment: ""), title: App.Title, okAction: {
                    let mapItem = placemarks?.first
                    self.zoomMapToLocation(mapItem!.placemark.location!)
                })
            }
            else { // Only one match found, so use it.
                let mapItem = placemarks?.first
                
                self.zoomMapToLocation(mapItem!.placemark.location!)
            }
        }
    }
    
    func showLoadingIndicator() {
        searchBtn.tintColor = UIColor(hexString: "#efefef")
        if !loadingIndicator!.isAnimating() {
            loadingIndicator.startAnimating()
        }
    }
    
    func hideLoadingIndicator() {
        searchBtn.tintColor = AppConfig.sharedInstance.theme?.primaryColor
        loadingIndicator.stopAnimating()
    }
    
    // MARK: - UITextFieldDelegate
    
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        
        if textField.text?.characters.count > 0 {
            searchLocationForAddress(textField.text!)
        }
        
        return true
    }
    
    // MARK: - MKMapViewDelegate
    
    func centerToLocation(location: CLLocation) {
        zoomMapToLocation(location)
    }
    
    func zoomMapToLocation(location: CLLocation) {
        if mapView == nil {
            return
        }
        
        var region = MKCoordinateRegion()
        region.center = location.coordinate
        region.span.latitudeDelta = 0.2
        region.span.longitudeDelta = 0.2
        mapView.setRegion(region, animated: true)
    }
    
    func mapView(mapView: MKMapView, didUpdateUserLocation userLocation: MKUserLocation) {
        if !showedFirstUserLocation {
            if toEdit == nil {
                centerToLocation(userLocation.location!)
            }
        }
        
        showedFirstUserLocation = true
    }
    
    func mapView(mapView: MKMapView, regionDidChangeAnimated animated: Bool) {
        if self.mapView == nil {
            return
        }
        
        showLoadingIndicator()
        
        let geoCoder = CLGeocoder()
        geoCoder.reverseGeocodeLocation(CLLocation(latitude: mapView.centerCoordinate.latitude, longitude: mapView.centerCoordinate.longitude)) { (placemarks, error) in
            self.hideLoadingIndicator()
            
            if error != nil {
                DDLogError("[\(NSStringFromClass(self.dynamicType))] Geocoder failed with error - \(error)")
                return
            }
            
            if placemarks != nil && placemarks?.count > 0 {
                self.lastPlacemark = placemarks?.first
                
                let addressText = ABCreateStringWithAddressDictionary(self.lastPlacemark!.addressDictionary!, false)
                let newString = addressText.componentsSeparatedByCharactersInSet(NSCharacterSet.newlineCharacterSet()).joinWithSeparator(", ")
                self.lastPlacemarkAddress = newString
            }
        }
    }
    
    // MARK: - Setup
    
    override func setupTopBar() {
        super.setupTopBar()
        
        UIApplication.sharedApplication().setStatusBarStyle(AppConfig.sharedInstance.theme!.mapStatusBarStyle!, animated: true)
    }
    
    func setupButtons() {
        let items = [backBtn, searchContainer]
        
        for view in items {
            Utils.applyShadowEffect(view, offset: CGSizeMake(1, 1), opacity: 0.1, radius: 2.0)
            view.layer.cornerRadius = 5.0
        }
        
        createBtn.backgroundColor = AppConfig.sharedInstance.theme?.primaryColor
        backBtn.tintColor = AppConfig.sharedInstance.theme?.primaryColor
    }
    
    func setupMapView() {
        mapView.rotateEnabled = false
        mapView.showsUserLocation = (CLLocationManager.authorizationStatus() == .AuthorizedAlways || CLLocationManager.authorizationStatus() == .AuthorizedWhenInUse)
    }
    
    func setupImageViews() {
        mapCenterImage.tintColor = AppConfig.sharedInstance.theme?.primaryColor
    }
}
